File: Sources/Olaf/UI/Util/HARExporter.swift
import Foundation

/// Görünen network kayıtlarını **HAR 1.2** (HTTP Archive) belgesine dönüştürür.
/// HAR; Charles, Proxyman, Chrome DevTools gibi araçlarca doğrudan açılır.
///
/// Kaynak, yakalama anındaki metadata'dır; birebir ölçülemeyen HAR alanları için spec'in
/// "bilinmiyor" değeri (`-1`) kullanılır. Gövdeler yakalandığı hâliyle yazılır (gerekirse
/// kesilmiş / pretty-print edilmiş). İçerik **ham**dır — paylaşmadan önce gözden geçirin.
enum HARExporter {

    /// Verilen kayıtlardan (network olmayanlar atlanır) HAR JSON metni üretir.
    /// Hiç network kaydı yoksa da geçerli (boş `entries`) bir belge döner.
    static func harDocument(from entries: [LogEntry]) -> String? {
        let harEntries = entries.compactMap { entry -> [String: Any]? in
            guard let info = NetworkLogInfo(entry: entry) else { return nil }
            return harEntry(info: info, date: entry.date)
        }
        let document: [String: Any] = [
            "log": [
                "version": "1.2",
                "creator": ["name": "Olaf", "version": "-"],
                "entries": harEntries
            ]
        ]
        guard let data = try? JSONSerialization.data(
            withJSONObject: document,
            options: [.prettyPrinted, .withoutEscapingSlashes, .sortedKeys]
        ) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    // MARK: - Entry

    private static func harEntry(info: NetworkLogInfo, date: Date) -> [String: Any] {
        let total = info.durationMs ?? 0
        // HAR sözleşmesi: `connect`, `ssl`'i kapsar → bilinen fazlar dns+connect+wait.
        let known = (info.dnsMs ?? 0) + (info.connectMs ?? 0) + (info.ttfbMs ?? 0)
        let receive = max(0, total - known)

        var request: [String: Any] = [
            "method": info.method ?? "GET",
            "url": info.url ?? "",
            "httpVersion": httpVersion(info.protocolName),
            "cookies": [] as [Any],
            "headers": harHeaders(info.requestHeaders),
            "queryString": queryString(of: info.url),
            "headersSize": -1,
            "bodySize": info.requestBytes ?? -1
        ]
        if let body = info.requestBody, !body.isEmpty {
            request["postData"] = [
                "mimeType": headerValue(info.requestHeaders, "Content-Type") ?? "application/octet-stream",
                "text": body
            ]
        }

        let response: [String: Any] = [
            "status": info.statusCode ?? 0,
            "statusText": info.statusCode.map { HTTPURLResponse.localizedString(forStatusCode: $0) }
                ?? (info.error ?? (info.cancelled ? "cancelled" : "")),
            "httpVersion": httpVersion(info.protocolName),
            "cookies": [] as [Any],
            "headers": harHeaders(info.responseHeaders),
            "content": [
                "size": info.responseBytes ?? -1,
                "mimeType": headerValue(info.responseHeaders, "Content-Type") ?? "application/octet-stream",
                "text": info.responseBody ?? ""
            ],
            "redirectURL": "",
            "headersSize": -1,
            "bodySize": info.responseBytes ?? -1
        ]

        return [
            "startedDateTime": Self.timestampFormatter.string(from: date),
            "time": total,
            "request": request,
            "response": response,
            "cache": [:] as [String: Any],
            "timings": [
                "blocked": -1,
                "dns": info.dnsMs ?? -1,
                "connect": info.connectMs ?? -1,
                "send": 0,
                "wait": info.ttfbMs ?? -1,
                "receive": receive,
                "ssl": info.tlsMs ?? -1
            ]
        ]
    }

    // MARK: - Yardımcılar

    private static func harHeaders(_ headers: [(key: String, value: String)]) -> [[String: String]] {
        headers.map { ["name": $0.key, "value": $0.value] }
    }

    private static func headerValue(_ headers: [(key: String, value: String)], _ name: String) -> String? {
        headers.first { $0.key.caseInsensitiveCompare(name) == .orderedSame }?.value
    }

    private static func queryString(of url: String?) -> [[String: String]] {
        guard let url, let items = URLComponents(string: url)?.queryItems else { return [] }
        return items.map { ["name": $0.name, "value": $0.value ?? ""] }
    }

    /// ALPN adını HAR'ın beklediği gösterime çevirir.
    private static func httpVersion(_ protocolName: String?) -> String {
        switch protocolName?.lowercased() {
        case "h2": return "HTTP/2"
        case "h3": return "HTTP/3"
        case let .some(name) where !name.isEmpty: return name
        default: return "HTTP/1.1"
        }
    }

    private static let timestampFormatter: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter
    }()
}

File: Android/olaf/src/main/java/com/olaf/ui/util/HarExporter.kt
package com.olaf.ui.util

import com.olaf.LogEntry
import com.olaf.ui.model.NetworkLogInfo
import org.json.JSONArray
import org.json.JSONObject
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

/**
 * Converts the visible network entries into a **HAR 1.2** document, which opens directly in
 * Charles, Proxyman or Chrome DevTools.
 *
 * The source is the metadata captured at request time; fields that cannot be measured exactly use
 * the spec's "unknown" value (`-1`). Bodies are written as captured — possibly truncated and
 * pretty-printed — and are **raw**, so review before sharing.
 */
internal object HarExporter {

    private val timestampFormatter: DateTimeFormatter =
        DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").withZone(ZoneOffset.UTC)

    /** Produces HAR JSON. Entries that aren't network records are skipped. */
    fun harDocument(entries: List<LogEntry>): String {
        val harEntries = JSONArray()
        entries.forEach { entry ->
            NetworkLogInfo.from(entry)?.let { harEntries.put(harEntry(it, entry.date)) }
        }

        val log = JSONObject()
            .put("version", "1.2")
            .put("creator", JSONObject().put("name", "Olaf").put("version", "-"))
            .put("entries", harEntries)

        return JSONObject().put("log", log).toString(2)
    }

    private fun harEntry(info: NetworkLogInfo, date: Instant): JSONObject {
        val total = info.durationMs ?: 0
        // Per the HAR contract `connect` already encompasses `ssl`, so the known phases are
        // dns + connect + wait; whatever is left over is time spent receiving.
        val known = (info.dnsMs ?: 0) + (info.connectMs ?: 0) + (info.ttfbMs ?: 0)
        val receive = (total - known).coerceAtLeast(0)

        val request = JSONObject()
            .put("method", info.method ?: "GET")
            .put("url", info.url.orEmpty())
            .put("httpVersion", httpVersion(info.protocolName))
            .put("cookies", JSONArray())
            .put("headers", harHeaders(info.requestHeaders))
            .put("queryString", queryString(info.url))
            .put("headersSize", -1)
            .put("bodySize", info.requestBytes ?: -1)

        info.requestBody?.takeIf { it.isNotEmpty() }?.let { body ->
            request.put(
                "postData",
                JSONObject()
                    .put("mimeType", headerValue(info.requestHeaders, "Content-Type") ?: "application/octet-stream")
                    .put("text", body)
            )
        }

        val response = JSONObject()
            .put("status", info.statusCode ?: 0)
            .put("statusText", info.error ?: if (info.cancelled) "cancelled" else "")
            .put("httpVersion", httpVersion(info.protocolName))
            .put("cookies", JSONArray())
            .put("headers", harHeaders(info.responseHeaders))
            .put(
                "content",
                JSONObject()
                    .put("size", info.responseBytes ?: -1)
                    .put("mimeType", headerValue(info.responseHeaders, "Content-Type") ?: "application/octet-stream")
                    .put("text", info.responseBody.orEmpty())
            )
            .put("redirectURL", "")
            .put("headersSize", -1)
            .put("bodySize", info.responseBytes ?: -1)

        val timings = JSONObject()
            .put("blocked", -1)
            .put("dns", info.dnsMs ?: -1)
            .put("connect", info.connectMs ?: -1)
            .put("send", 0)
            .put("wait", info.ttfbMs ?: -1)
            .put("receive", receive)
            .put("ssl", info.tlsMs ?: -1)

        return JSONObject()
            .put("startedDateTime", timestampFormatter.format(date))
            .put("time", total)
            .put("request", request)
            .put("response", response)
            .put("cache", JSONObject())
            .put("timings", timings)
    }

    private fun harHeaders(headers: List<Pair<String, String>>): JSONArray {
        val array = JSONArray()
        headers.forEach { (name, value) ->
            array.put(JSONObject().put("name", name).put("value", value))
        }
        return array
    }

    private fun headerValue(headers: List<Pair<String, String>>, name: String): String? =
        headers.firstOrNull { it.first.equals(name, ignoreCase = true) }?.second

    private fun queryString(url: String?): JSONArray {
        val array = JSONArray()
        val query = url?.substringAfter('?', "").orEmpty()
        if (query.isEmpty()) return array
        query.split('&').forEach { pair ->
            if (pair.isEmpty()) return@forEach
            array.put(
                JSONObject()
                    .put("name", pair.substringBefore('='))
                    .put("value", pair.substringAfter('=', ""))
            )
        }
        return array
    }

    /** Maps the ALPN name onto the representation HAR expects. */
    private fun httpVersion(protocolName: String?): String = when (protocolName?.lowercase()) {
        "h2" -> "HTTP/2"
        "h3" -> "HTTP/3"
        null, "" -> "HTTP/1.1"
        else -> protocolName
    }
}
